File: internal/store/detail.go
package store

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"os"
)

// CompanyDetail is the payload for GET /api/companies/:id.
// Fields are flattened and JSON-tagged for direct serialization.
type CompanyDetail struct {
	CompanyID    string            `json:"company_id"`
	Name         string            `json:"name"`
	Source       string            `json:"source"`
	SourceID     string            `json:"source_id"`
	Domain       string            `json:"domain"`
	Headcount    int64             `json:"headcount"`
	FundingStage string            `json:"funding_stage"`
	Location     string            `json:"location"`
	Vertical     string            `json:"vertical"`
	IngestedAt   string            `json:"ingested_at"`
	RawJSON      map[string]string `json:"raw_json"`

	HasVerdict   bool   `json:"has_verdict"`
	Verdict      string `json:"verdict"`
	Reason       string `json:"reason"`
	TasteVersion string `json:"taste_version"`
	Model        string `json:"model"`
	ScoredAt     string `json:"scored_at"`

	HasEnrichment  bool   `json:"has_enrichment"`
	WebsiteURL     string `json:"website_url"`
	WebsiteSummary string `json:"website_summary"`
	FetchStatus    string `json:"fetch_status"`
	FetchError     string `json:"fetch_error"`
	FetchedAt      string `json:"fetched_at"`

	Postings []Posting `json:"postings"`
}

// GetCompanyDetail returns the full joined detail for one company.
// Returns nil, nil if not found.
func (db *DB) GetCompanyDetail(companyID string) (*CompanyDetail, error) {
	const q = `
SELECT c.id, c.name, c.source, COALESCE(c.source_id, ''),
       COALESCE(c.domain, ''), COALESCE(c.headcount, 0),
       COALESCE(c.funding_stage, ''), COALESCE(c.location, ''),
       COALESCE(c.vertical, ''), c.ingested_at, c.raw_json,
       v.verdict, v.reason, v.taste_version, v.model, v.scored_at,
       e.website_url, e.website_summary, e.fetch_status, e.fetch_error, e.fetched_at
FROM companies c
LEFT JOIN verdicts   v ON v.company_id = c.id
LEFT JOIN enrichment e ON e.company_id = c.id
WHERE c.id = ?`

	var d CompanyDetail
	var rawJSON string
	var verdict, reason, tasteVersion, model, scoredAt sql.NullString
	var websiteURL, websiteSummary, fetchStatus, fetchError, fetchedAt sql.NullString

	err := db.QueryRow(q, companyID).Scan(
		&d.CompanyID, &d.Name, &d.Source, &d.SourceID,
		&d.Domain, &d.Headcount, &d.FundingStage, &d.Location, &d.Vertical,
		&d.IngestedAt, &rawJSON,
		&verdict, &reason, &tasteVersion, &model, &scoredAt,
		&websiteURL, &websiteSummary, &fetchStatus, &fetchError, &fetchedAt,
	)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	d.RawJSON = parseRawJSON(rawJSON)
	if verdict.Valid {
		d.HasVerdict = true
		d.Verdict = verdict.String
		d.Reason = reason.String
		d.TasteVersion = tasteVersion.String
		d.Model = model.String
		d.ScoredAt = scoredAt.String
	}
	if websiteURL.Valid || websiteSummary.Valid || fetchStatus.Valid {
		d.HasEnrichment = true
		d.WebsiteURL = websiteURL.String
		d.WebsiteSummary = websiteSummary.String
		d.FetchStatus = fetchStatus.String
		d.FetchError = fetchError.String
		d.FetchedAt = fetchedAt.String
	}

	// Postings are one-to-many, so they ride a second query rather than the
	// join above. A failure here shouldn't sink the whole detail payload, but
	// log it so a real DB error isn't silently masked by an empty slice.
	if postings, err := db.ListPostings(companyID); err == nil {
		d.Postings = postings
	} else {
		fmt.Fprintf(os.Stderr, "list postings %s: %v\n", companyID, err)
		d.Postings = []Posting{}
	}

	return &d, nil
}

// parseRawJSON turns the stored raw_json column (an object of header->value)
// into a map. We tolerate any shape on disk — if it doesn't parse cleanly,
// the caller gets an empty map.
func parseRawJSON(s string) map[string]string {
	out := map[string]string{}
	if s == "" {
		return out
	}
	// Try object-of-strings first (the ingest shape).
	var m1 map[string]string
	if err := json.Unmarshal([]byte(s), &m1); err == nil {
		return m1
	}
	// Fallback: any object, stringify values.
	var m2 map[string]any
	if err := json.Unmarshal([]byte(s), &m2); err == nil {
		for k, v := range m2 {
			out[k] = fmt.Sprintf("%v", v)
		}
	}
	return out
}

// Stats is the payload for GET /api/stats.
type Stats struct {
	TotalCompanies    int            `json:"total_companies"`
	EnrichedOK        int            `json:"enriched_ok"`
	Scored            int            `json:"scored"`
	Unscored          int            `json:"unscored"`
	ByVerdict         map[string]int `json:"by_verdict"`
	FetchStatus       map[string]int `json:"fetch_status"`
	CurrentTaste      string         `json:"current_taste"`       // version hash, e.g. "b4cd783174d6"
	TasteSource       string         `json:"taste_source"`        // "file:taste.md" or "brainbot:<url>" or "" if unknown
	StaleVerdicts     int            `json:"stale_verdicts"`      // verdicts whose taste_version != CurrentTaste
	TasteVersionsSeen []string       `json:"taste_versions_seen"` // distinct taste_versions present in verdicts
}

// GetStats computes the sidebar payload. currentTasteVersion may be empty
// (e.g. taste.md unreadable at serve start); when empty, StaleVerdicts is 0.
func (db *DB) GetStats(currentTasteVersion, currentTasteSource string) (*Stats, error) {
	s := &Stats{
		ByVerdict:    map[string]int{},
		FetchStatus:  map[string]int{},
		CurrentTaste: currentTasteVersion,
		TasteSource:  currentTasteSource,
	}

	if err := db.QueryRow(`SELECT COUNT(1) FROM companies`).Scan(&s.TotalCompanies); err != nil {
		return nil, err
	}
	if err := db.QueryRow(`SELECT COUNT(1) FROM enrichment WHERE fetch_status = 'ok'`).Scan(&s.EnrichedOK); err != nil {
		return nil, err
	}
	if err := db.QueryRow(`SELECT COUNT(1) FROM verdicts`).Scan(&s.Scored); err != nil {
		return nil, err
	}
	s.Unscored = s.TotalCompanies - s.Scored
	if s.Unscored < 0 {
		s.Unscored = 0
	}

	if err := scanHist(db, `SELECT verdict, COUNT(1) FROM verdicts GROUP BY verdict`, s.ByVerdict); err != nil {
		return nil, err
	}
	if err := scanHist(db, `SELECT fetch_status, COUNT(1) FROM enrichment GROUP BY fetch_status`, s.FetchStatus); err != nil {
		return nil, err
	}

	rows, err := db.Query(`SELECT DISTINCT taste_version FROM verdicts ORDER BY taste_version`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	for rows.Next() {
		var v string
		if err := rows.Scan(&v); err != nil {
			return nil, err
		}
		s.TasteVersionsSeen = append(s.TasteVersionsSeen, v)
		if currentTasteVersion != "" && v != currentTasteVersion {
			// Count stale below in a single query for accuracy; we'll overwrite.
		}
	}
	if currentTasteVersion != "" {
		if err := db.QueryRow(
			`SELECT COUNT(1) FROM verdicts WHERE taste_version != ?`,
			currentTasteVersion,
		).Scan(&s.StaleVerdicts); err != nil {
			return nil, err
		}
	}

	return s, nil
}

func scanHist(db *DB, q string, dst map[string]int) error {
	rows, err := db.Query(q)
	if err != nil {
		return err
	}
	defer rows.Close()
	for rows.Next() {
		var k string
		var n int
		if err := rows.Scan(&k, &n); err != nil {
			return err
		}
		dst[k] = n
	}
	return rows.Err()
}

// GetCompanyName looks up name + domain by ID. Used by the brain proxy route.
func (db *DB) GetCompanyName(companyID string) (name, domain string, err error) {
	err = db.QueryRow(
		`SELECT name, COALESCE(domain, '') FROM companies WHERE id = ?`,
		companyID,
	).Scan(&name, &domain)
	if err == sql.ErrNoRows {
		return "", "", sql.ErrNoRows
	}
	return name, domain, err
}

File: scout/store/detail.py
"""Joined company-detail + stats payloads."""

from __future__ import annotations

import json
import sqlite3
import sys
from dataclasses import dataclass, field

from . import errors, postings, taste_filter


@dataclass
class CompanyDetail:
    company_id: str = ""
    name: str = ""
    source: str = ""
    source_id: str = ""
    domain: str = ""
    headcount: int = 0
    funding_stage: str = ""
    location: str = ""
    vertical: str = ""
    ingested_at: str = ""
    raw_json: dict = field(default_factory=dict)
    notes: str = ""
    flagged: bool = False
    flagged_at: str = ""
    reviewed_at: str = ""
    has_verdict: bool = False
    verdict: str = ""
    reason: str = ""
    taste_version: str = ""
    model: str = ""
    scored_at: str = ""
    has_enrichment: bool = False
    website_url: str = ""
    website_summary: str = ""
    fetch_status: str = ""
    fetch_error: str = ""
    fetched_at: str = ""
    postings: list = field(default_factory=list)


def get_company_detail(con: sqlite3.Connection, company_id: str) -> CompanyDetail | None:
    """The full joined detail for one company, or None when not found."""
    q = """
SELECT c.id, c.name, c.source, COALESCE(c.source_id, ''),
       COALESCE(c.domain, ''), COALESCE(c.headcount, 0),
       COALESCE(c.funding_stage, ''), COALESCE(c.location, ''),
       COALESCE(c.vertical, ''), c.ingested_at, c.raw_json, c.flagged_at, c.reviewed_at,
       COALESCE(c.notes, ''),
       v.verdict, v.reason, v.taste_version, v.model, v.scored_at,
       e.website_url, e.website_summary, e.fetch_status, e.fetch_error, e.fetched_at
FROM companies c
LEFT JOIN verdicts   v ON v.company_id = c.id
LEFT JOIN enrichment e ON e.company_id = c.id
WHERE c.id = ?"""
    row = con.execute(q, (company_id,)).fetchone()
    if row is None:
        return None

    d = CompanyDetail(
        company_id=row[0],
        name=row[1],
        source=row[2],
        source_id=row[3],
        domain=row[4],
        headcount=row[5],
        funding_stage=row[6],
        location=row[7],
        vertical=row[8],
        ingested_at=row[9],
        notes=row[13],
    )
    d.raw_json = _parse_raw_json(row[10])
    flagged_at, reviewed_at = row[11], row[12]
    d.flagged = flagged_at is not None
    d.flagged_at = flagged_at or ""
    d.reviewed_at = reviewed_at or ""

    verdict, reason, taste_version, model, scored_at = row[14], row[15], row[16], row[17], row[18]
    if verdict is not None:
        d.has_verdict = True
        d.verdict = verdict
        d.reason = reason or ""
        d.taste_version = taste_version or ""
        d.model = model or ""
        d.scored_at = scored_at or ""

    website_url, website_summary, fetch_status = row[19], row[20], row[21]
    fetch_error, fetched_at = row[22], row[23]
    if website_url is not None or website_summary is not None or fetch_status is not None:
        d.has_enrichment = True
        d.website_url = website_url or ""
        d.website_summary = website_summary or ""
        d.fetch_status = fetch_status or ""
        d.fetch_error = fetch_error or ""
        d.fetched_at = fetched_at or ""

    # Postings are one-to-many; a failure here shouldn't sink the whole payload.
    try:
        d.postings = postings.list_postings(con, company_id)
    except Exception as exc:  # noqa: BLE001 - log and continue
        print(f"list postings {company_id}: {exc}", file=sys.stderr)
        d.postings = []
    return d


def _parse_raw_json(s) -> dict:
    """Turn the stored raw_json column into a map; tolerate any on-disk shape."""
    out: dict = {}
    if not s:
        return out
    try:
        m = json.loads(s)
    except (json.JSONDecodeError, TypeError, ValueError):
        return out
    if isinstance(m, dict):
        if all(isinstance(v, str) for v in m.values()):
            return m
        return {k: _stringify(v) for k, v in m.items()}
    return out


def _stringify(v) -> str:
    if isinstance(v, bool):
        return "true" if v else "false"
    if v is None:
        return "<nil>"
    return str(v)


@dataclass
class Stats:
    total_companies: int = 0
    enriched_ok: int = 0
    scored: int = 0
    unscored: int = 0
    by_verdict: dict = field(default_factory=dict)
    fetch_status: dict = field(default_factory=dict)
    current_taste: str = ""
    taste_source: str = ""
    taste_filter_enabled: bool = True


def get_stats(
    con: sqlite3.Connection, current_taste_version: str, current_taste_source: str
) -> Stats:
    """Compute the sidebar payload."""
    s = Stats(
        by_verdict={},
        fetch_status={},
        current_taste=current_taste_version,
        taste_source=current_taste_source,
        taste_filter_enabled=True,
    )
    try:
        _, enabled = taste_filter.get_taste_filter(con)
        s.taste_filter_enabled = enabled
    except Exception:  # noqa: BLE001
        pass

    s.total_companies = con.execute("SELECT COUNT(1) FROM companies").fetchone()[0]
    s.enriched_ok = con.execute(
        "SELECT COUNT(1) FROM enrichment WHERE fetch_status = 'ok'"
    ).fetchone()[0]
    s.scored = con.execute("SELECT COUNT(1) FROM verdicts").fetchone()[0]
    s.unscored = max(s.total_companies - s.scored, 0)

    _scan_hist(con, "SELECT verdict, COUNT(1) FROM verdicts GROUP BY verdict", s.by_verdict)
    _scan_hist(
        con, "SELECT fetch_status, COUNT(1) FROM enrichment GROUP BY fetch_status", s.fetch_status
    )
    return s


def _scan_hist(con: sqlite3.Connection, q: str, dst: dict) -> None:
    for k, n in con.execute(q).fetchall():
        dst[k] = n


def get_company_name(con: sqlite3.Connection, company_id: str) -> tuple[str, str]:
    """Look up (name, domain) by id. Raises NotFound when absent."""
    row = con.execute(
        "SELECT name, COALESCE(domain, '') FROM companies WHERE id = ?", (company_id,)
    ).fetchone()
    if row is None:
        raise errors.NotFound()
    return row[0], row[1]
